mq/inmemory/mq.h

```diff
@@ -77,72 +164,103 @@ class MMQ final {
-            return;  // LCOV_EXCL_LINE
-          }
-          condition_variable_.wait(lock, [this] { return head_ready_ != tail_ || destructing_; });
+        while (circular_buffer_[tail].status != Entry::READY) {
           if (destructing_) {
             return;
           }
+          condition_variable_.wait(
+              lock, [this, tail] { return (circular_buffer_[tail].status == Entry::READY) || destructing_; });
+        }
+        if (destructing_) {
+          return;
         }
-        this_time_dropped_messages = number_of_dropped_messages_;
-        number_of_dropped_messages_ = 0;
+        circular_buffer_[tail].status = Entry::BEING_EXPORTED;
+
+        actually_dropped_messages = dropped_messages_count_;
+        dropped_messages_count_ = 0u;
       }
 
       {
         // Then, export the message.
         // NO MUTEX REQUIRED.
-        consumer_.OnMessage(std::ref(circular_buffer_[index].message_body), this_time_dropped_messages);
+        ExportMessage(consumer_, std::move(circular_buffer_[tail].message_body), actually_dropped_messages);
       }
 
       {
-        // Finally, mark the message as successfully exported.
+        // Finally, mark the message entry in the buffer as `FREE` for overwriting.
         // MUTEX-LOCKED.
-        std::lock_guard<std::mutex> lock(mutex_);
-        Increment(tail_);
+        {
+          std::lock_guard<std::mutex> lock(mutex_);
+          circular_buffer_[tail].status = Entry::FREE;
+        }
+        Increment(tail);
+
+        // Need to notify message pushers.
+        // TODO(dkorolev) + TODO(mzhurovich): Think whether this might be a performance bottleneck.
+        condition_variable_.notify_one();
       }
     }
   }
 
-  size_t PushMessageAllocate() {
-    // First, allocate room in the buffer for this message.
-    // Overwrite the oldest message if have to.
+  template <bool DROP = DROP_ON_OVERFLOW>
+  typename std::enable_if<DROP, size_t>::type PushMessageAllocate() {
+    // Implementation that discards message if the queue is full.
     // MUTEX-LOCKED.
     std::lock_guard<std::mutex> lock(mutex_);
-    const size_t index = head_allocated_;
-    Increment(head_allocated_);
-    if (head_allocated_ == tail_) {
-      // Buffer overflow, must drop the least recent element and keep the count of those.
-      ++number_of_dropped_messages_;
-      if (tail_ == head_ready_) {
-        Increment(head_ready_);  // LCOV_EXCL_LINE
+    if (circular_buffer_[head_].status == Entry::FREE) {
+      // Regular case.
+      const size_t index = head_;
+      Increment(head_);
+      circular_buffer_[index].status = Entry::BEING_IMPORTED;
+      return index;
+    } else {
+      // Overflow. Discarding the message.
+      return static_cast<size_t>(-1);
+    }
+  }
+
+  template <bool DROP = DROP_ON_OVERFLOW>
+  typename std::enable_if<!DROP, size_t>::type PushMessageAllocate() {
+    // Implementation that waits for empty space if the queue is full.
```

Comment:
`for an empty space`, and I'd note that this call blocks the calling thread, potentially indefinitely.